slack_integration: Use the new `UrlOptions.MAPPING` options.

PieterCK opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented on zerver/webhooks/slack/doc.md:
```diff
@@ -27,19 +27,6 @@ If you are looking to quickly move your Slack integrations to Zulip, check out
-
-    To send all Slack messages into a **single Zulip topic**, enable the **Send
-    all notifications to a single topic** option, with no further modifications.
-
```
I feel like current UX makes these configs feel intuitive, but I also know all the context surrounding the config. What instructions or info do we want to leave here, if any?

PieterCK commented:
@zulipbot add "buddy review"

@Aditya8840 Do you have the time to do a pass on this?

Aditya8840 commented:
> @zulipbot add "buddy review"
> 
> @Aditya8840 Do you have time to do a pass on this?

Yeah, looking into!

PieterCK commented:
Updated the icons for the "Matcing Zulip channels" to `icon-hash` and "Matcing Zulip topics" to `icon-chevron-rigt` as per discussed in CZO

PieterCK commented:
@zulipbot add "mentor review"
@chdinesh1089 -- Would appreciate you reviewing this if you have a chance

PieterCK commented:
> channels_map_to_topics=1: "topics" - Choose a single channel name from stream dropdown, leave "single topic" option unchecked.

Aha, I thought something was off. Thanks for checking this! Indeed, this is the new behavior we want to replace the `channels_map_to_topics=1` option with, not sure why I had gone down the road of adding another `UrlOption`.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -116,6 +116,24 @@ export function show_generate_integration_url_modal(api_key: string): void {
+                $checkbox_element.prop("checked", is_checked);
+            }
+        }
+
```
The main reason I included the toggle for checking or unchecking is because we have one chase for that -- when one selects the "Matching Zulip channel" a checkbox is disabled but checked automatically. I think I'll stick with your suggestion since this is a much better improvement readability wise!

PieterCK commented:
@Niloth-p Appreciate the quick and thorough review! I'll get to this tomorrow :+1:

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -421,15 +451,27 @@ export function show_generate_integration_url_modal(api_key: string): void {
+                toggle_input_checkbox_property($map_to_topics_checkbox, true, false);
             } else {
-                toggle_input_checkbox_property($override_topic, false, undefined);
+                if ($map_to_topics_checkbox.prop("checked")) {
+                    toggle_input_checkbox_property($override_topic, true, undefined);
+                } else {
+                    toggle_input_checkbox_property($override_topic, false, undefined);
+                }
+                if ($override_topic.prop("checked")) {
+                    toggle_input_checkbox_property($map_to_topics_checkbox, true, undefined);
+                } else {
+                    toggle_input_checkbox_property($map_to_topics_checkbox, false, undefined);
+                }
```
Ah, I think I used if-else here because the linter was complaining `is_disabled` accepts `boolean` but  `.prop()` returns `any`. 

Updated to use `.is()` instead:
```suggestion
                 set_checkbox_disabled_state(
                    $override_topic,
                    $map_to_topics_checkbox.is(":checked"),
                );
                set_checkbox_disabled_state(
                    $map_to_topics_checkbox,
                    $override_topic.is(":checked"),
                );
```

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -185,6 +203,18 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            const checked = this.checked;
+            // This setting and $map_to_topics_checkbox can be off at the same time
+            // but only one can be selected.
+            $map_to_topics_checkbox = get_map_to_topics_checkbox_element();
```
Yeah, I'm not really sure what's the naming convention or pattern is for this, I just thought it'd be clearer if I add the `_checkbox` suffix. We do use the `_checkbox` suffix for checkboxes in other modules like stream_settings, stream_create, etc.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -185,6 +203,18 @@ export function show_generate_integration_url_modal(api_key: string): void {
             }
         });
 
+        $override_topic.on("click", function () {
+            const checked = this.checked;
+            // This setting and $map_to_topics_checkbox can be off at the same time
+            // but only one can be selected.
+            $map_to_topics_checkbox = get_map_to_topics_checkbox_element();
+            if (checked) {
+                toggle_input_checkbox_property($map_to_topics_checkbox, true, false);
+            } else {
+                toggle_input_checkbox_property($map_to_topics_checkbox, false, false);
+            }
+        });
```
That's neat! Updated to this

PieterCK commented:
@Niloth-p -- OK! I think I've addressed your comments, marking this as ready for you to take a look again. Other than the changes that we discussed in the comment threads, I added a new commit to rename `MAPPING` to `CHANNELS_MAPPING`: 
- `integrations: Rename url_option MAPPING to CHANNEL_MAPPING.`

I've also manually tested all the cases listed in the PR description again.

PieterCK commented:
@zulipbot remove "mentor review"

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -103,6 +104,12 @@ def build_preset_config(cls, config: PresetUrlOption) -> "WebhookUrlOption":
+            case PresetUrlOption.MAP_TO_TOPICS:  # nocoverage # Not used yet
+                return cls(
+                    name=config.value,
+                    label="Send messages to topics with the same name as the original channel?",
```
Hmm, I wonder if it's still worth keeping the string as is for external services that use the "channel"(e.g., MS Teams uses "channel")? That's an easy thing to tweak, though, I'll update it to "Send messages to topics named after Slack channels" for now!

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -167,6 +171,19 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            $map_to_topics_checkbox.add($override_topic).on("click", function () {
+                const other_checkbox =
+                    this === $override_topic[0] ? $map_to_topics_checkbox : $override_topic;
+                $map_to_topics_checkbox = get_map_to_topics_checkbox_element();
```
Turns out this isn't necessary, removed it

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -277,9 +318,32 @@ def api_slack_webhook(
+    mapping_option: str | None = map_to_topics or map_to_channels
+
+    requires_channel_name = (channels_map_to_topics or mapping_option) is not None
+    if requires_channel_name:
+        channel = get_slack_channel_name(channel_id, slack_app_token)
+    else:
+        channel = None
```
Updated to the suggested change!

PieterCK commented:
@Niloth-p Thanks for doing another pass on this! Turns out the bug was because I was missing a couple of lines here:
```diff
@@ -440,6 +439,8 @@ export function show_generate_integration_url_modal(api_key: string): void {
             } else if (user_selected_option === map_channels_option.unique_id) {
                 $override_topic.prop("checked", true);
                 set_checkbox_disabled_state($override_topic, true);
+                $map_to_topics_checkbox.prop("checked", false);
+                set_checkbox_disabled_state($map_to_topics_checkbox, true);
                 $topic_input.val("");
                 $topic_input.parent().removeClass("hide");
             } else {
```
I've updated the screen recordings and the screenshot in the PR description.

>Looks like your editor is currently inserting trailing commas everywhere instead of just for multiline cases.

This was not the editor's fault, actually :sweat_smile:. I think I had to do that once in a PR, but I couldn't recall where. Probably the function had very verbose parameters but still within 3 parameters. Anywho, I've removed the commas for non-multi line cases.

alya commented:
If you leave the topic blank, will messages go to *general chat*? If so, we should be sure to allow that, and also hint that this will happen with placeholder text:

> Leave blank for *general chat*

I expect we aren't checking channel permissions for topics in this modal (whether general chat is allowed/required)? I don't think we necessarily have to, but might be a fine follow-up to note if it's doable.

PieterCK commented:
hmm I haven't come up with a clean solution for the custom "topic to use" dropdown field yet. It's doable but will probably be messy since there'll be some special casing, I'll have to play around for a bit to see if there's a nice implementation for this.

>If you leave the topic blank, will messages go to general chat? -

The existing behavior for black topic is to send the messages to a topic called "Message from Slack". I think the only problem with changing this behavior is that we'll lose backward compatibility. i.e., folks using existing integrations might still expect that their notifications are sent to the "Message from Slack" topic. what do you think?

alya commented:
Hm, that's a pretty weird default, but yeah, let's consider changing it a separate issue/discussion. For now, though, the UI should (a) tell you what's gonna happen, and (b) make it possible to send to _general chat_.

PieterCK commented:
Okay, I think a setup that'll allow for both backwards compatibility and the option to send to _general chat_ is probably something like this:

| UI | Behavior |
|--------|--------|
| <img width="809" height="672" alt="image" src="https://github.com/user-attachments/assets/55c1f3f9-cfdd-4a50-9ce5-a711f78d799d" /> | Not selecting any "Topics to use" options will default to the existing behavior, send notifications to the  "Message from Slack" topic. This is `channels_map_to_topics=None` (not specified). |
| <img width="809" height="785" alt="image" src="https://github.com/user-attachments/assets/958b7e37-d66a-4ecd-8779-fd4ad9a2d19c" />| Selecting "Send all notifications to a single topic" option but leaving the "Topic" field empty will send notifications to the _general chat_ topic. |
| <img width="809" height="669" alt="image" src="https://github.com/user-attachments/assets/03cc7cd0-afd2-47ac-8910-918be72b76db" /> | The `channels_map_to_topics=1` config is ported to the "Topics named after Slack channels" option | 

Differences from Alya's suggestions (https://github.com/zulip/zulip/pull/35680#issuecomment-3348282491 & https://github.com/zulip/zulip/pull/35680#issuecomment-3353722049):
- **The proposed change**: Use "Message from Slack" as the default value of the "Topic" field to preserve existing behavior. **This version**: if none of the options from "Topics to use" is selected, use the default behavior. I think this is a simpler approach, both UX wise and technical wise (don't have to do extra rewiring to preserve backwards compatibility).
- **The proposed change**: In the "Topics to use" dropdown, the "Custom string" option shows the "Topic" text field. **This version:** I think using the original string -- "Send all notifications to a single topic" -- would be more accurate than "Custom topic". Leaving the "Topic" field empty sends notifications to _general chat_.

What do folks think?

alya commented:
I'd appreciate @timabbott 's feedback here, but in general, I don't want anything to happen silently, without the user seeing in the UI what will occur. So, if "Message from Slack" is the topic to be used, it would be good for users to see that in a topic field in this modal. It's fine to disable the "Copy URL" button when the configuration is incomplete, to hint that there's more work to be done.

alya commented:
@Niloth-p Is this one ready for @timabbott 's review from your perspective?

Niloth-p commented on web/src/integration_url_modal.ts:
```diff
@@ -242,10 +260,24 @@ export function show_generate_integration_url_modal(api_key: string): void {
                 params.set("stream", stream_id!.toString());
-                if ($override_topic.prop("checked") && topic_name !== "") {
+
+                // The topics_to_use_dropdown_widget is only used by the Slack integration,
```
And once we rename the variables, we can drop this comment (just this line) too - "only used by the Slack integration".

Niloth-p commented on web/src/integration_url_modal.ts:
```diff
@@ -359,6 +396,42 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            },
+            item_click_callback: topics_to_use_click_callback,
+            $events_container: $("#generate-integration-url-modal"),
+            default_id: default_topics_to_use_option.unique_id,
```
```suggestion
	        text_if_current_value_not_in_options: $t({
	            defaultMessage: "Select notification topics",
	        }),
```

Niloth-p commented on web/src/integration_url_modal.ts:
```diff
@@ -359,6 +396,42 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            widget_name: topics_to_use_widget_id,
+            get_options() {
+                return [
+                    default_topics_to_use_option,
```
Drop this from the selectable options.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -59,19 +59,37 @@ export function show_generate_integration_url_modal(api_key: string): void {
         const $dialog_submit_button = $("#generate-integration-url-modal .dialog_submit_button");
         const $show_integration_events = $("#show-integration-events");
         const $config_container = $("#integration-url-config-options-container");
+        const $topics_to_use_widget = $(`#${topics_to_use_widget_id}_widget`);
```
renamed to:
```js
        const slack_topics_dropdown_widget_id = "slack-topics-dropdown";

        const $slack_topics_dropdown = $(`#${slack_topics_dropdown_widget_id}_widget`);
```

PieterCK commented:
I think this is ready for @Niloth-p to take a look again!. Main changes:
- Renamed various variable names from "topic to use" naming to "slack topics dropdown" naming.
- Added `set_topic_config_field_state` to manage the state of the "topics to use" fields (`$override_topic` and `$slack_topics_dropdown`). Previously these two fields was manually toggled, making bugs caused by inconsistent state between the two fields easy to happen.
- Used `text_if_current_value_not_in_options` instead of a selectable default option.
- Programmatically switching to a specific `DropdownWidget` option from the state where `text_if_current_value_not_in_options` is selected by calling `DropdownWidget.render(some_option)` was impossible, added this commit to fix this behavior `dropdown_widget: Fix buggy value check when calling render().`
- Switching to the custom text state was also impossible, added `reset_selected_option_to_custom_text` to do that.

```diff
diff --git a/web/src/dropdown_widget.ts b/web/src/dropdown_widget.ts
index 8bdd9379f5..2806a02e23 100644
--- a/web/src/dropdown_widget.ts
+++ b/web/src/dropdown_widget.ts
@@ -712,7 +712,12 @@ export class DropdownWidget {
     // dropdowns don't need it. Maybe we can follow a reverse approach in the future.
     render(value?: number | string): void {
         // Check if the value is valid otherwise just render previous value.
-        if (value !== undefined && typeof value === typeof this.current_value) {
+        const current_value_type = this.text_if_current_value_not_in_options
+            ? String(this.unique_id_type)
+            : typeof this.current_value;
+
+        // eslint-disable-next-line valid-typeof
+        if (value !== undefined && typeof value === current_value_type) {
             this.current_value = value;
         }

@@ -747,4 +752,16 @@ export class DropdownWidget {
             $(this.widget_value_selector).text(option.name);
         }
     }
+
+    reset_selected_option_to_custom_text(): void {
+        if (this.text_if_current_value_not_in_options) {
+            this.current_value = undefined;
+            $(this.widget_value_selector).html(
+                render_dropdown_current_value_not_in_options({
+                    name: this.text_if_current_value_not_in_options,
+                }),
+            );
+            return;
+        }
+    }
 }
diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
index ccb26da0f4..87892ef613 100644
--- a/web/src/integration_url_modal.ts
+++ b/web/src/integration_url_modal.ts
...skipping...
+                send_all_to_single_field ? send_all_to_single_topic_option.unique_id : undefined,
+            );
+            $override_topic.prop("checked", send_all_to_single_field);
+
+            if (!send_all_to_single_field) {
+                slack_topics_dropdown_widget.reset_selected_option_to_custom_text();
+            }
+        }
+
         function render_url_options(config: UrlOption[]): void {
             const validated_config = url_options_schema.parse(config);
             $config_container.empty();
@@ -260,24 +278,24 @@ export function show_generate_integration_url_modal(api_key: string): void {
             }

             const params = new URLSearchParams({api_key});
-            const selected_topic_option = topics_to_use_dropdown_widget.value();
+            const selected_slack_topic_option = slack_topics_dropdown_widget.value();
             if (stream_id !== -1) {
                 params.set("stream", stream_id!.toString());

-                // The topics_to_use_dropdown_widget is only used by the Slack integration,
-                // it replaces the $override_topic field.
+                // Only used by the Slack integration, it replaces the $override_topic field.
                 const send_notification_to_a_single_topic: boolean =
                     Boolean($override_topic.prop("checked")) ||
-                    selected_topic_option === send_all_to_single_topic_option.unique_id;
+                    selected_slack_topic_option === send_all_to_single_topic_option.unique_id;

-                if (send_notification_to_a_single_topic && topic_name !== "") {
-                    params.set("topic", topic_name);
-                } else if (
-                    selected_topic_option === topics_named_after_slack_channels_option.unique_id
+                if (
+                    selected_slack_topic_option ===
+                    topics_named_after_slack_channels_option.unique_id
                 ) {
-                    params.delete("topic");
                     // This setting is from the Slack webhook integration.
                     params.set("channels_map_to_topics", "1");
+                } else if (send_notification_to_a_single_topic && topic_name !== "") {
+                    params.delete("channels_map_to_topics");
+                    params.set("topic", topic_name);
                 }
             }

@@ -372,10 +390,10 @@ export function show_generate_integration_url_modal(api_key: string): void {
             );
             const url_options = selected_integration_data?.url_options;
             if (selected_integration === "slack") {
-                $override_topic.closest(".input-group").toggleClass("hide", true);
-                topics_to_use_dropdown_widget.setup();
-                set_input_disabled_state($topics_to_use_widget, true);
-                $topics_to_use_widget.closest(".input-group").toggleClass("hide", false);
+                $override_topic.closest(".input-group").hide();
+                slack_topics_dropdown_widget.setup();
+                set_input_disabled_state($slack_topics_dropdown, true);
+                $slack_topics_dropdown.closest(".input-group").show();
             }
             if (url_options) {
                 render_url_options(url_options);
@@ -396,35 +414,34 @@ export function show_generate_integration_url_modal(api_key: string): void {
         });
         stream_input_dropdown_widget.setup();

-        $topics_to_use_widget.closest(".input-group").toggleClass("hide", true);
-        topics_to_use_dropdown_widget = new dropdown_widget.DropdownWidget({
-            widget_name: topics_to_use_widget_id,
+        $slack_topics_dropdown.closest(".input-group").hide();
+        slack_topics_dropdown_widget = new dropdown_widget.DropdownWidget({
+            widget_name: slack_topics_dropdown_widget_id,
             get_options() {
-                return [
-                    default_topics_to_use_option,
-                    topics_named_after_slack_channels_option,
-                    send_all_to_single_topic_option,
-                ];
+                return [topics_named_after_slack_channels_option, send_all_to_single_topic_option];
             },
-            item_click_callback: topics_to_use_click_callback,
+            item_click_callback: slack_topics_click_callback,
             $events_container: $("#generate-integration-url-modal"),
-            default_id: default_topics_to_use_option.unique_id,
+            text_if_current_value_not_in_options: $t({
+                defaultMessage: "Select notification topics",
+            }),
             unique_id_type: "string",
         });

-        function topics_to_use_click_callback(
+        function slack_topics_click_callback(
             event: JQuery.ClickEvent,
             dropdown: tippy.Instance,
+            widget: DropdownWidget,
         ): void {
-            topics_to_use_dropdown_widget.render();
-            $topics_to_use_widget.trigger("input");
+            slack_topics_dropdown_widget.render();
+            $slack_topics_dropdown.trigger("input");
+            const selected_slack_topic_option = widget.value()!;

-            const selected_topic_option = topics_to_use_dropdown_widget.value();
             $topic_input
                 .parent()
                 .toggleClass(
                     "hide",
-                    selected_topic_option !== send_all_to_single_topic_option.unique_id,
+                    selected_slack_topic_option !== send_all_to_single_topic_option.unique_id,
                 );

             dropdown.hide();
@@ -483,21 +500,14 @@ export function show_generate_integration_url_modal(api_key: string): void {
             dropdown.hide();
             const user_selected_option = stream_input_dropdown_widget.value();
             if (user_selected_option === direct_messages_option.unique_id) {
-                $override_topic.prop("checked", false);
-                set_input_disabled_state($override_topic, true);
+                set_topic_config_field_state(true, false);
                 $topic_input.val("");
-                set_input_disabled_state($topics_to_use_widget, true);
-                topics_to_use_dropdown_widget.render(default_topics_to_use_option.unique_id);
             } else if (user_selected_option === map_channels_option.unique_id) {
-                $override_topic.prop("checked", true);
-                set_input_disabled_state($override_topic, true);
+                set_topic_config_field_state(true, true);
                 $topic_input.val("");
                 $topic_input.parent().removeClass("hide");
-                set_input_disabled_state($topics_to_use_widget, true);
-                topics_to_use_dropdown_widget.render(send_all_to_single_topic_option.unique_id);
             } else {
-                set_input_disabled_state($override_topic, false);
-                set_input_disabled_state($topics_to_use_widget, false);
+                set_topic_config_field_state(false);
             }
             $override_topic.trigger("change");
             event.preventDefault();
@@ -531,15 +541,15 @@ export function show_generate_integration_url_modal(api_key: string): void {
             $("#integrations-event-container .integration-event").prop("checked", false);
             $show_integration_events.prop("checked", false);

-            $topics_to_use_widget.closest(".input-group").toggleClass("hide", true);
-            $override_topic.closest(".input-group").toggleClass("hide", false);
-            $("#integration-url-override-topic").toggleClass("hide", false);
-            set_input_disabled_state($override_topic, true);
+            $slack_topics_dropdown.closest(".input-group").hide();
+            $override_topic.closest(".input-group").show();
+            $("#integration-url-override-topic").show();
+            set_topic_config_field_state(true);
             $topic_input.val("");
             $topic_input.parent().addClass("hide");

             stream_input_dropdown_widget.render(direct_messages_option.unique_id);
-            topics_to_use_dropdown_widget.render(default_topics_to_use_option.unique_id);
+            slack_topics_dropdown_widget.render();
             $config_container.empty();
             branch_pill_widget = undefined;
         }
diff --git a/web/templates/settings/generate_integration_url_modal.hbs b/web/templates/settings/generate_integration_url_modal.hbs
index 2db20a7a0d..69e43d0de4 100644
--- a/web/templates/settings/generate_integration_url_modal.hbs
+++ b/web/templates/settings/generate_integration_url_modal.hbs
@@ -24,7 +24,7 @@
 <div class="input-group">
     <div class="topic-to-use-wrapper integration-url-parameter">
         {{> ../dropdown_widget_with_label
-          widget_name="topics-to-use"
+          widget_name="slack-topics-dropdown"
           label=(t "Topics to use")}}
     </div>
 </div>
```

Niloth-p commented on web/src/dropdown_widget.ts:
```diff
@@ -712,7 +712,12 @@ export class DropdownWidget {
     render(value?: number | string): void {
         // Check if the value is valid otherwise just render previous value.
-        if (value !== undefined && typeof value === typeof this.current_value) {
+        const current_value_type = this.text_if_current_value_not_in_options
```
I'm not sure if this is a bug or if it's explicitly kept this way to discourage calling render with a string.
I made the mistake of passing a string directly to render() in another PR. We want to uniformly use Option objects to keep them translatable.

And yes, currently `setup()` doesn't call `render()` when the current value is `undefined`. Sahil Batra recommended setting `default_id` to "" for that. Would that be sufficient here?

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -111,6 +124,36 @@ export function show_generate_integration_url_modal(api_key: string): void {
+
+        function set_topic_config_field_state(
+            disabled: boolean,
+            send_all_to_single_field?: boolean,
```
Oof, seems like a busted rename I didn't bother to re-check. How about `set_to_send_all_to_a_single_topic?` or `select_send_all_to_a_single_topic_option?`

PieterCK replied:
Renamed the variable name to `select_send_all_to_a_single_topic_option`:
```diff
diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
index 49f0231ec7..8fcc31c82b 100644
--- a/web/src/integration_url_modal.ts
+++ b/web/src/integration_url_modal.ts
         function set_topic_config_field_state(
-            disabled: boolean,
-            send_all_to_single_field?: boolean,
+            disable: boolean,
+            select_send_all_to_a_single_topic_option?: boolean,
         ): void {
             // The topic configuration uses different UI depending on the integration.
             // The Slack integration uses a dropdown field while other integrations use
             // a checkbox field.
-            set_input_disabled_state($override_topic, disabled);
-            set_input_disabled_state($slack_topics_dropdown, disabled);
+            set_input_disabled_state($override_topic, disable);
+            set_input_disabled_state($slack_topics_dropdown, disable);

-            if (send_all_to_single_field === undefined) {
+            if (select_send_all_to_a_single_topic_option === undefined) {
                 return;
             }
             slack_topics_dropdown_widget.render(
-                send_all_to_single_field ? send_all_to_single_topic_option.unique_id : undefined,
+                select_send_all_to_a_single_topic_option
+                    ? send_all_to_single_topic_option.unique_id
+                    : undefined,
             );
-            $override_topic.prop("checked", send_all_to_single_field);
+            $override_topic.prop("checked", select_send_all_to_a_single_topic_option);

-            if (!send_all_to_single_field) {
+            if (!select_send_all_to_a_single_topic_option) {
                 slack_topics_dropdown_widget.reset_selected_option_to_custom_text();
             }
         }
```

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -111,6 +124,36 @@ export function show_generate_integration_url_modal(api_key: string): void {
+        }
+
+        function set_topic_config_field_state(
+            disabled: boolean,
```
Updated to `disable`

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -111,6 +124,36 @@ export function show_generate_integration_url_modal(api_key: string): void {
             update_url();
         }
 
+        function set_input_disabled_state($checkbox: JQuery, disable: boolean): void {
```
Renamed the parameter name:
```diff
diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
index 49f0231ec7..8fcc31c82b 100644
--- a/web/src/integration_url_modal.ts
+++ b/web/src/integration_url_modal.ts
@@ -124,32 +124,34 @@ export function show_generate_integration_url_modal(api_key: string): void {
             update_url();
         }

-        function set_input_disabled_state($checkbox: JQuery, disable: boolean): void {
-            $checkbox
+        function set_input_disabled_state($input: JQuery, disable: boolean): void {
+            $input
                 .prop("disabled", disable)
                 .closest(".input-group")
                 .toggleClass("control-label-disabled", disable);
         }
```

PieterCK commented:
> There's another bug: The "topics to use" dropdown is not getting reset properly. On value updates from both the integration dropdown and the stream dropdown.

- Fix for the bug where the "topics to use" dropdown is not getting reset properly on value update from the integration dropdown:
  ```diff
  diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
  index 49f0231ec7..8fcc31c82b 100644
  --- a/web/src/integration_url_modal.ts
  +++ b/web/src/integration_url_modal.ts
  
  @@ -382,7 +384,7 @@ export function show_generate_integration_url_modal(api_key: string): void {
               if (selected_integration === "slack") {
                   $override_topic.closest(".input-group").hide();
                   slack_topics_dropdown_widget.setup();
  -                set_input_disabled_state($slack_topics_dropdown, true);
  +                set_topic_config_field_state(true, false);
                   $slack_topics_dropdown.closest(".input-group").show();
               }
               if (url_options) {
  ```
  Previously, if Slack is selected, it will only disable the "topics to use" field and not reset its value. Updated to use `set_topic_config_field_state` to thoroughly reset it. `stream_item_click_callback` will reset `$topic_input`.

- Fix for the bug where the "topics to use" dropdown is not getting reset properly on value update from the channel dropdown:
  ```diff
  diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
  index 49f0231ec7..8fcc31c82b 100644
  --- a/web/src/integration_url_modal.ts
  +++ b/web/src/integration_url_modal.ts
  @@ -497,7 +500,8 @@ export function show_generate_integration_url_modal(api_key: string): void {
                   $topic_input.val("");
                   $topic_input.parent().removeClass("hide");
               } else {
  -                set_topic_config_field_state(false);
  +                set_topic_config_field_state(false, false);
  +                $topic_input.val("");
               }
               $override_topic.trigger("change");
               event.preventDefault();
  ```
  This seems like a new rule to me, since I think in `main` we don't touch the value of "Send all notifications to a single topic" or even the "Topic" on value change from channel dropdown (?) Or did I misunderstand the bug?

Added a couple more screencaps of manual tests in the description for these bugs!

Niloth-p commented on web/src/integration_url_modal.ts:
```diff
@@ -111,6 +124,38 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            if (!select_send_all_to_a_single_topic_option) {
+                slack_topics_dropdown_widget.reset_selected_option_to_custom_text();
+            }
+        }
```
```suggestion
        function update_topic_ui({
            disable_topic_config_inputs,
            show_topic_input,
        }: TopicUIParams): void {
            set_input_disabled_state($override_topic, disable_topic_config_inputs);
            set_input_disabled_state($slack_topics_dropdown, disable_topic_config_inputs);

            if (show_topic_input === undefined) {
                return;
            }

            $override_topic.prop("checked", show_topic_input);
            if (show_topic_input) {
                slack_topics_dropdown_widget.render(send_all_to_single_topic_option.unique_id);
            } else {
                slack_topics_dropdown_widget.reset_selected_option_to_custom_text();
            }
        }
```

- Use named parameters so that the function calls are readable, when we're passing two boolean values that are not obvious from the function name.
- Use a single if-else instead of double checking.
- No need to call render and then reset for the false case. Directly call reset.
- Rename `set_topic_config_field_state` function to something simpler like `update_topic_ui`.
- Rename `disable` to `disable_topic_config_inputs` or something like that so that it's readable. I thought of that since your comment mentions "Topic configuration UI".
- Rename `select_send_all_to_a_single_topic_option` to `use_single_topic` or `show_topic_input` - that still conveys the purpose. `show_topic_input` would make the function much more readable `update_topic_ui({disable_topic_config_ui, show_topic_input})`.
- Move the comment to where we define TopicUIParams. And re-write it like below.

```ts
type TopicUIParams = {
    // The topic configuration inputs are:
    // override_topic and Slack-specific slack_topics_dropdown_widget
    disable_topic_config_inputs: boolean;
    show_topic_input?: boolean;
};
```

The names and comment are just suggestions, feel free to come up with better ones in the same direction.

PieterCK replied:
Those changes looks good to me! Updated to:

```diff
diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
index b186c84e89..4a60f68edb 100644
--- a/web/src/integration_url_modal.ts
+++ b/web/src/integration_url_modal.ts
@@ -25,6 +25,13 @@ type UrlOption = {
     validator: string;
 };

+type TopicUIParams = {
+    // The topic configuration inputs are:
+    // override_topic and Slack-specific slack_topics_dropdown_widget
+    disable_topic_config_inputs: boolean;
+    show_topic_input?: boolean;
+};
+
 const url_option_schema = z.object({
     key: z.string(),
     label: z.string(),
@@ -131,27 +138,19 @@ export function show_generate_integration_url_modal(api_key: string): void {
                 .toggleClass("control-label-disabled", disable);
         }

-        function set_topic_config_field_state(
-            disable: boolean,
-            select_send_all_to_a_single_topic_option?: boolean,
-        ): void {
-            // The topic configuration uses different UI depending on the integration.
-            // The Slack integration uses a dropdown field while other integrations use
-            // a checkbox field.
-            set_input_disabled_state($override_topic, disable);
-            set_input_disabled_state($slack_topics_dropdown, disable);
-
-            if (select_send_all_to_a_single_topic_option === undefined) {
+        function update_topic_ui({
+            disable_topic_config_inputs,
+            show_topic_input,
+        }: TopicUIParams): void {
+            set_input_disabled_state($override_topic, disable_topic_config_inputs);
+            set_input_disabled_state($slack_topics_dropdown, disable_topic_config_inputs);
+            if (show_topic_input === undefined) {
                 return;
             }
+            
-            slack_topics_dropdown_widget.render(
-                select_send_all_to_a_single_topic_option
-                    ? send_all_to_single_topic_option.unique_id
-                    : undefined,
-            );
-            $override_topic.prop("checked", select_send_all_to_a_single_topic_option);
-
-            if (!select_send_all_to_a_single_topic_option) {
+            
+            $override_topic.prop("checked", show_topic_input);
+            if (show_topic_input) {
+                slack_topics_dropdown_widget.render(send_all_to_single_topic_option.unique_id);
+            } else {
                 slack_topics_dropdown_widget.reset_selected_option_to_custom_text();
             }
         }
@@ -384,7 +383,10 @@ export function show_generate_integration_url_modal(api_key: string): void {
             if (selected_integration === "slack") {
                 $override_topic.closest(".input-group").hide();
                 slack_topics_dropdown_widget.setup();
-                set_topic_config_field_state(true, false);
+                update_topic_ui({
+                    disable_topic_config_inputs: true,
+                    show_topic_input: false,
+                });
                 $slack_topics_dropdown.closest(".input-group").show();
             }
             if (url_options) {
@@ -489,14 +491,22 @@ export function show_generate_integration_url_modal(api_key: string): void {
             dropdown.hide();
             const user_selected_option = stream_input_dropdown_widget.value();
             if (user_selected_option === direct_messages_option.unique_id) {
-                set_topic_config_field_state(true, false);
+                update_topic_ui({
+                    disable_topic_config_inputs: true,
+                    show_topic_input: false,
+                });
                 $topic_input.val("");
             } else if (user_selected_option === map_channels_option.unique_id) {
-                set_topic_config_field_state(true, true);
+                update_topic_ui({
+                    disable_topic_config_inputs: true,
+                    show_topic_input: true,
+                });
                 $topic_input.val("");
                 $topic_input.parent().removeClass("hide");
             } else {
-                set_topic_config_field_state(false, undefined);
+                update_topic_ui({
+                    disable_topic_config_inputs: false,
+                });
             }
             $override_topic.trigger("change");
             event.preventDefault();
@@ -533,7 +543,9 @@ export function show_generate_integration_url_modal(api_key: string): void {
             $slack_topics_dropdown.closest(".input-group").hide();
             $override_topic.closest(".input-group").show();
             $("#integration-url-override-topic").show();
-            set_topic_config_field_state(true);
+            update_topic_ui({
+                disable_topic_config_inputs: true,
+            });
             $topic_input.val("");
             $topic_input.parent().addClass("hide");
```

Niloth-p reviewed (commented):
The screen captures in the PR description are pretty thorough, thanks!
I cannot review `modal.css`, someone else should.
Apart from the below refactor, everything LGTM!
You can tag this for integration review once you're done with the refactor. :rocket:

PieterCK commented:
Pushed [these](https://github.com/zulip/zulip/pull/35680#discussion_r2649570651) changes and manually tested again.

@Niloth-p Thanks! So glad this got another set of reviews, ironed out quite a handful of bugs from these changes.

@zulipbot remove "maintainer review"

sahil839 commented on web/src/integration_url_modal.ts:
```diff
@@ -329,7 +382,15 @@ export function show_generate_integration_url_modal(api_key: string): void {
-
+            if (selected_integration === "slack") {
+                $override_topic.closest(".input-group").hide();
+                slack_topics_dropdown_widget.setup();
```
I think `generate_integration_url_post_render` has become too large and a good follow-up project might to atleast extract functions for setting up the integrations and this dropdown.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -349,6 +410,38 @@ export function show_generate_integration_url_modal(api_key: string): void {
+                "checked",
+                widget.value() === send_all_to_single_topic_option.unique_id,
+            );
+            $override_topic.trigger("change");
```
- Refactored a `topic_ui_callback`, which can be used by both topic UIs
```diff
@@ -200,12 +198,16 @@ export function show_generate_integration_url_modal(api_key: string): void {
             }
         }

-        $override_topic.on("change", function () {
-            const checked = this.checked;
-            $topic_input.parent().toggleClass("hide", !checked);
+        function topic_ui_callback(show_topic_text_field: boolean): void {
+            $topic_input.parent().toggleClass("hide", !show_topic_text_field);
             if (!$topic_input.parent().hasClass("hide")) {
                 $topic_input.trigger("focus");
             }
+        }
+
+        $override_topic.on("change", function () {
+            const checked = this.checked;
+            topic_ui_callback(checked);
         });

         $show_integration_events.on("change", () => {
@@ -432,11 +438,7 @@ export function show_generate_integration_url_modal(api_key: string): void {
         ): void {
             slack_topics_dropdown_widget.render();
             $slack_topics_dropdown.trigger("input");
-            $override_topic.prop(
-                "checked",
-                widget.value() === send_all_to_single_topic_option.unique_id,
-            );
-            $override_topic.trigger("change");
+            topic_ui_callback(widget.value() === send_all_to_single_topic_option.unique_id);
             dropdown.hide();
             event.preventDefault();
             event.stopPropagation();
@@ -510,7 +512,15 @@ export function show_generate_integration_url_modal(api_key: string): void {
                     disable_topic_config_inputs: false,
                 });
             }
-            $override_topic.trigger("change");
+
+            if (selected_integration === "slack") {
+                topic_ui_callback(
+                    slack_topics_dropdown_widget.value() ===
+                        send_all_to_single_topic_option.unique_id,
+                );
+            } else {
+                $override_topic.trigger("change");
+            }
             event.preventDefault();
             event.stopPropagation();
         }
```

PieterCK commented on web/src/dropdown_widget.ts:
```diff
@@ -756,4 +756,16 @@ export class DropdownWidget {
         }
     }
+
+    reset_selected_option_to_custom_text(): void {
```
Yep, `reset_selected_option_to_custom_text` is not needed, dropped it and called `render("")` instead:
```diff
diff --git a/web/src/dropdown_widget.ts b/web/src/dropdown_widget.ts
index 21c8cc0838..0696af1b2f 100644
--- a/web/src/dropdown_widget.ts
+++ b/web/src/dropdown_widget.ts
@@ -775,16 +775,4 @@ export class DropdownWidget {
             $(this.widget_value_selector).text(option.name);
         }
     }
-
-    reset_selected_option_to_custom_text(): void {
-        if (this.text_if_current_value_not_in_options) {
-            this.current_value = this.default_id;
-            $(this.widget_value_selector).html(
-                render_dropdown_current_value_not_in_options({
-                    name: this.text_if_current_value_not_in_options,
-                }),
-            );
-            return;
-        }
-    }
 }
diff --git a/web/src/integration_url_modal.ts b/web/src/integration_url_modal.ts
index 5048938c66..6cc3434c95 100644
--- a/web/src/integration_url_modal.ts
+++ b/web/src/integration_url_modal.ts
@@ -150,11 +150,9 @@ export function show_generate_integration_url_modal(api_key: string): void {
             }

             $override_topic.prop("checked", show_topic_input);
-            if (show_topic_input) {
-                slack_topics_dropdown_widget.render(send_all_to_single_topic_option.unique_id);
-            } else {
-                slack_topics_dropdown_widget.reset_selected_option_to_custom_text();
-            }
+            slack_topics_dropdown_widget.render(
+                show_topic_input ? send_all_to_single_topic_option.unique_id : "",
+            );
         }

         function render_url_options(config: UrlOption[]): void {
```

PieterCK commented:
@sahil839 -- Thanks for reviewing this, I think I've addressed all of your comments, and re-tested the various scenarios listed in the description after that. Marking this as ready for review again.

sahil839 reviewed (commented):
Posted a couple of comments. Looks good otherwise.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -349,6 +417,35 @@ export function show_generate_integration_url_modal(api_key: string): void {
+            widget: DropdownWidget,
+        ): void {
+            slack_topics_dropdown_widget.render();
+            $slack_topics_dropdown.trigger("input");
```
Yeah, this isn't doing anything. Removed it

sahil839 reviewed (commented):
Posted one comment. Looks good otherwise.

PieterCK commented:
@sahil839 -- Thanks for doing a pass on this! This is ready for review again
